Update CONTRIBUTING.md: add Backwards Compatibility guide and update questions channels (#1304)
use-ink/ink: CONTRIBUTING.md

   This might seem pedantic but we believe that in essence this is going to improve overall comment and documentation quality.
1. If possible try to sign your commits, e.g. using GPG keys. For more information about this go [here](https://help.github.com/en/articles/signing-commits).

### Backwards Compatibility

> ink! and pallet_contracts are the projects under active development. As contracts API functions evolve to their newer versions, we need to introduce them in ink! in a way that still keeps current *MAJOR* ink! version compatible with elder substrate nodes built with the pallet_contracts version having not this new functionality.

Reviewer: ```suggestion
ink! and pallet_contracts are the projects under active development. As contracts API functions evolve to their newer versions, we need to introduce them in ink! in a way that still keeps current *MAJOR* ink! version compatible with older substrate nodes built with the pallet_contracts version without this new functionality.
```

> ink! and pallet_contracts are the projects under active development. As contracts API functions evolve to their newer versions, we need to introduce them in ink! in a way that still keeps current *MAJOR* ink! version compatible with elder substrate nodes built with the pallet_contracts version having not this new functionality.

Reviewer 2: ```suggestion
ink! and pallet_contracts are the projects under active development. As contracts API functions (denoted by the `seal_*` prefix) evolve to their newer versions, we need to introduce them in ink! in a way that still keeps current *MAJOR* ink! version compatible with older Substrate nodes. These older nodes may have a version of the Contracts pallet which does not support new seal APIs.
```

Author: left it to be Andrew's variant for the sake of concise

In order to achieve this, please stick to the following workflow.

> Imagine there is a `[seal0] function()` in pallet\_contracts API and our `ink_env::function()` uses its import under the hood.

Reviewer 2: ```suggestion
Imagine there is a `[seal0] function()` in the Contract pallet API and our `ink_env::function()` imports it under the hood.
```

Author: well function does not import, it uses an imported one

> Then some time later we may decide to add a new `[seal1] function()` version to the pallet. In order to introduce it into ink! and still be able to run contracts (which don't use this particular function) on elder nodes, please do the following:

Reviewer: ```suggestion
Then some time later we may decide to add a new `[seal1] function()` version to the pallet. In order to introduce it into ink! and still be able to run contracts (which don't use this particular function) on older nodes, please do the following:
```

1. Mark the old `ink_env::function()` (which depends on the imported `[seal0]` function) with the `#[deprecated]` attribute.  
  Please, specify the `since` field with the ink! version since which this function is deprecated in favor of the newest one, and is left there only for the backwards compatibility purposes. Specifing some additional helpful info in the `note` field could also be a good idea.

> 2. Name the new function (which depends on the `[seal1] function()`) somehow descriptive like `ink_env::function_whats_special()`.

Reviewer 2: Mark all of these as `1`, the markdown render will handle the numbering

Author: well, this is a optional feature, btw it will work so even if you mark it with a randomly assigned set of integer numbers as well. However, I think it's good to have it numbered properly in the source itself.

> 3. Always add the `# Compatibility` section to the docs comment of the new function.

Reviewer 2: ```suggestion
3. Add a `# Compatibility` section to the docs comment of the new function.
```

> 4. Never use the new function with existing language features. Only with newly added ones.

Reviewer 2: ```suggestion
4. Never use the new function with existing language features. Only use it with newly added functions.
```
